nrf-hal-common/src/i2s.rs

```diff
@@ -0,0 +1,580 @@
+        self
+    }
+
+    /// Sets the frame format.
+    #[inline(always)]
+    pub fn set_format(&self, format: Format) -> &Self {
+        self.i2s
+            .config
+            .format
+            .write(|w| w.format().bit(format.into()));
+        self
+    }
+
+    /// Sets the I2S channel configuation.
+    #[inline(always)]
+    pub fn set_channels(&self, channels: Channels) -> &Self {
+        self.i2s
+            .config
+            .channels
+            .write(|w| unsafe { w.channels().bits(channels.into()) });
+        self
+    }
+
+    /// Sets the transmit data buffer (TX).
+    /// NOTE: The TX buffer must live until the transfer is done, or corrupted data will be transmitted.
+    #[inline(always)]
+    pub fn tx_buffer<B: I2SBuffer + ?Sized>(&self, buf: &B) -> Result<(), Error> {
+        if (buf.ptr() as usize) < SRAM_LOWER || (buf.ptr() as usize) > SRAM_UPPER {
+            return Err(Error::DMABufferNotInDataMemory);
+        }
+
+        if buf.maxcnt() > MAX_DMA_MAXCNT {
+            return Err(Error::BufferTooLong);
+        }
+
+        self.i2s
+            .txd
+            .ptr
+            .write(|w| unsafe { w.ptr().bits(buf.ptr()) });
+        self.i2s
+            .rxtxd
+            .maxcnt
+            .write(|w| unsafe { w.bits(buf.maxcnt()) });
+
+        Ok(())
+    }
+
+    /// Sets the receive data buffer (RX).
+    #[inline(always)]
+    pub fn rx_buffer<B: I2SBuffer + ?Sized>(&self, buf: &'static mut B) -> Result<(), Error> {
+        if (buf.ptr() as usize) < SRAM_LOWER || (buf.ptr() as usize) > SRAM_UPPER {
+            return Err(Error::DMABufferNotInDataMemory);
+        }
+
+        if buf.maxcnt() > MAX_DMA_MAXCNT {
+            return Err(Error::BufferTooLong);
+        }
+
+        self.i2s
+            .rxd
+            .ptr
+            .write(|w| unsafe { w.ptr().bits(buf.ptr()) });
+        self.i2s
+            .rxtxd
+            .maxcnt
+            .write(|w| unsafe { w.bits(buf.maxcnt()) });
+
+        Ok(())
+    }
+
+    /// Sets the transmit buffer RAM start address.
+    #[inline(always)]
+    pub fn set_tx_ptr(&self, addr: u32) -> Result<(), Error> {
+        if (addr as usize) < SRAM_LOWER || (addr as usize) > SRAM_UPPER {
+            return Err(Error::DMABufferNotInDataMemory);
+        }
+        self.i2s.txd.ptr.write(|w| unsafe { w.ptr().bits(addr) });
+        Ok(())
+    }
+
+    /// Sets the receive buffer RAM start address.
+    #[inline(always)]
+    pub unsafe fn set_rx_ptr(&self, addr: u32) -> Result<(), Error> {
+        if (addr as usize) < SRAM_LOWER || (addr as usize) > SRAM_UPPER {
+            return Err(Error::DMABufferNotInDataMemory);
+        }
+        self.i2s.rxd.ptr.write(|w| w.ptr().bits(addr));
+        Ok(())
+    }
+
+    /// Sets the size (in 32bit words) of the receive and transmit buffers.
+    #[inline(always)]
+    pub fn set_buffersize(&self, n_32bit: u32) -> Result<(), Error> {
```

Comment:
This one's also unsafe when used with `rx_buffer` since it writes an arbitrary value to `maxcnt`.

Reply:
True! :)